IOS-4203 IOS-4175 Update implementation sending parameters for analytics

skibinalexander opened the pull request:
Исправил передачу понятно url для аналитики. Поэтому пришлось добавить префиксы. И изменил проверку MultinetworkProvider. Старался делать так , чтобы не сильно менялись контракты

skibinalexander committed:
IOS-4203 IOS-4175 Update implementation sending parameters for analytics

skibinalexander committed:
IOS-4175 Fix current host value

skibinalexander committed:
IOS-4175 Compare only host value

skibinalexander committed:
IOS-4175 Check verify host parameter for switch

skibinalexander committed:
IOS-4175 Merge branch 'develop' into IOS-4175-snap-switch-api-host

skibinalexander commented:
Пока не закидывал в ревью чтобы посмотреть имплементацию и собрать сборку для QA

tureck1y commented on BlockchainSdk/BlockchainService/BlockBook/Config/Implementations/GetBlockBlockBookConfig.swift:
```diff
@@ -9,9 +9,13 @@
     private let apiKey: String
     
-    init(apiKey: String) {
+    private let basicHost = "getblock.io"
```
надо наоборот, не блокчейн тащить в GetBlockBlockBookConfig,
а в BlockBookUtxoProvider его дополнять. Там вот такой код

    var host: String {
        config.host
    }

и знание о блокчейне есть

skibinalexander replied:
Да, хорошее замечание, переделал + выдернул знание о блокчейне совсем

tureck1y commented on BlockchainSdk/BlockchainService/BlockBook/Config/Implementations/NowNodesBlockBookConfig.swift:
```diff
@@ -9,14 +9,19 @@
-    init(apiKey: String) {
+    private let basicHost = "nownodes.io"
+    
+    init(currencySymbol: String, apiKey: String) {
```
тут то же самое соответственно

tureck1y commented on BlockchainSdk/BlockchainService/Blockcypher/BlockcypherNetworkProvider.swift:
```diff
@@ -15,7 +15,7 @@ import TangemSdk
     var supportsTransactionPush: Bool { false }
     var host: String {
-        getTarget(for: .fee).baseURL.hostOrUnknown
+        return getTarget(for: .fee).baseURL.absoluteString
```
return не пишем в var

skibinalexander replied:
Поправил

tureck1y commented on BlockchainSdk/Common/API/MultiNetworkProvider.swift:
```diff
@@ -65,12 +67,17 @@ extension MultiNetworkProvider {
     // in the code, don't forget to update also Solano.Swift framework, class NetworkingRouter
     private func switchProviderIfNeeded(for errorHost: String) -> String? {
-        if errorHost != self.host { // Do not switch the provider, if it was switched already
+        if errorHost.hostOrNil != self.host.hostOrNil { // Do not switch the provider, if it was switched already and hostOrNil compare only host value
```
Вроде hostOrNil не нужен, потому что сравнивать надо полностью по строке в host. Например мы ходим к некоторым провайдером с капи ключом и без. В твоем случае будет баг и не переключится

skibinalexander committed:
IOS-4175 Merge branch 'develop' into IOS-4175-snap-switch-api-host

skibinalexander committed:
IOS-4175 Merge branch 'develop' into IOS-4175-snap-switch-api-host

skibinalexander committed:
IOS-4175 Fix review comments

skibinalexander commented:
Еще протестирую, пока перевьюить изменения можно

skibinalexander commented on BlockchainSdkExample/BlockchainSdkExampleViewModel.swift:
```diff
@@ -409,7 +409,7 @@ class BlockchainSdkExampleViewModel: ObservableObject {
     
     private func createWalletManager(blockchain: Blockchain, wallet: Card.Wallet) throws -> WalletManager {
-        let publicKey = Wallet.PublicKey(seedKey: wallet.publicKey, derivation: .none)
+        let publicKey = Wallet.PublicKey(seedKey: wallet.publicKey, derivationType: .none)
```
Фикс, потому что падает в девелопе, уйдет видимо когда соберемся заливать, иначе не протестировать

skibinalexander committed:
IOS-4175 Fix revert comment

megakoko commented on BlockchainSdk/Common/API/MultiNetworkProvider.swift:
```diff
@@ -42,11 +41,14 @@ extension MultiNetworkProvider {
                     // Send event if api did switched by host value
+                    
                     if currentHost != nextHost {
-                        Log.network("Switching to next publisher on host")
-                        
+                        Log.network("Switching to next publisher on host \(nextHost)")
+
```
Лишние изменения пробелов

skibinalexander replied:
Хотя стоп почему лишние?

megakoko replied:
Почему две строчки подсвечиваются если менял одну?

skibinalexander replied:
Добавил чтобы не слипалось

skibinalexander replied:
> Что не слипалось? Тут пробелы удалены
> 
> <img alt="Screenshot 2023-08-24 at 14 05 20" width="297" src="https://user-images.githubusercontent.com/12209839/262953609-f3fe4e1a-dcd2-452e-9e4f-bdbe39b97287.png">

Ревертнул

megakoko commented on BlockchainSdk/BlockchainService/BlockBook/Config/Implementations/NowNodesBlockBookConfig.swift:
```diff
@@ -17,6 +17,7 @@ struct NowNodesBlockBookConfig {
 }
 
 extension NowNodesBlockBookConfig: BlockBookConfig {
+    
```
Лишнее

megakoko commented on BlockchainSdk/BlockchainService/Blockcypher/BlockcypherNetworkProvider.swift:
```diff
@@ -15,7 +15,7 @@ import TangemSdk
     var supportsTransactionPush: Bool { false }
     var host: String {
-        getTarget(for: .fee).baseURL.hostOrUnknown
+        getTarget(for: .fee).baseURL.absoluteString
```
А зачем это изменения?

skibinalexander replied:
От какого блокчейна

megakoko commented on BlockchainSdk/BlockchainService/BlockBook/BlockBookUtxoProvider.swift:
```diff
@@ -12,7 +12,7 @@ import Combine
 class BlockBookUtxoProvider {
     var host: String {
-        config.host
+        "\(blockchain.currencySymbol).\(config.host)"
```
Давай сделаем config.domain, там то же самое

skibinalexander replied:
> Давай сделаем config.domain, там то же самое

Коммент от @tureck1y "не тащим блокчейн в конфиг, а дополняем его здесь,  в виду наличия инфы"

megakoko replied:
Уверен, что Саша это имел в виду?

skibinalexander replied:
https://github.com/tangem/blockchain-sdk-swift/pull/389#discussion_r1296071796

tureck1y replied:
Domain не подойдет. Нужен стабильный айди хоста, не зависящий от запроса

megakoko commented on BlockchainSdk/BlockchainService/BlockBook/Config/Implementations/GetBlockBlockBookConfig.swift:
```diff
@@ -29,14 +29,12 @@ extension GetBlockBlockBookConfig: BlockBookConfig {
-        
+    func domain(for request: BlockBookTarget.Request, prefix: String, isTestnet: Bool) -> String {
         switch request {
         case .fees:
-            return "https://\(currencySymbolPrefix).\(host)"
+            return "https://\(prefix).\(host)"
         default:
-            return "https://\(currencySymbolPrefix).\(host)"
+            return "https://\(prefix).\(host)"
         }
     }
```
А зачем эти изменения?

skibinalexander replied:
> А зачем эти изменения?

По комменту выше, не тащить сюда блокчейн

tureck1y replied:
Надо убрать вообще функцию domain for. Лепи префикс к хосту в потребителе и все. Зачем такие сложности.

tureck1y replied:
Эти получаются лишние в рамках задачм

tureck1y replied:
Я бы откатил их, в остальном ок

skibinalexander replied:
Да, окей откачусь

skibinalexander replied:
Зарезолвилось при мердже последних изменений из develop

skibinalexander committed:
IOS-4175 Add lowercased BlockBookUtxoProvider

skibinalexander committed:
IOS-4175 Merge branch 'develop' into IOS-4175-snap-switch-api-host

# Conflicts:
#	BlockchainSdk/BlockchainService/BlockBook/BlockBookTarget.swift
#	BlockchainSdk/BlockchainService/BlockBook/Config/BlockBookConfig.swift
#	BlockchainSdk/BlockchainService/BlockBook/Config/Implementations/GetBlockBlockBookConfig.swift
#	BlockchainSdk/BlockchainService/BlockBook/Config/Implementations/NowNodesBlockBookConfig.swift

skibinalexander committed:
IOS-4175 Fix review comments

skibinalexander committed:
IOS-4175 Fix review comment by multinetwork provider
